### ISSUE TITLE
Internal Server Error when trying to delete a book

### ISSUE THREAD

**Issue by H4L35** [janeczku/calibre-web#1810]
When I go to Books list, and click the delete button, I get an error (below), the book remains, although the majority of the metadata seems to have been removed. 

Running on Ubuntu

I get the error:

Calibre-Web
Internal Server Error
500 Internal Server Error: The server encountered an internal error and was unable to complete your request. Either the server is overloaded or there is an error in the application.
Traceback (most recent call last):
File "/home/seedit4me/calibre-web/vendor/sqlalchemy/engine/base.py", line 1277, in _execute_context
cursor, statement, parameters, context
File "/home/seedit4me/calibre-web/vendor/sqlalchemy/engine/default.py", line 609, in do_execute
cursor.execute(statement, parameters)
sqlite3.OperationalError: no such module: fts5
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
File "/home/seedit4me/calibre-web/cps/editbooks.py", line 281, in delete_book
calibre_db.session.query(db.Books).filter(db.Books.id == book_id).delete()
File "/home/seedit4me/calibre-web/vendor/sqlalchemy/orm/query.py", line 3926, in delete
delete_op.exec_()
File "/home/seedit4me/calibre-web/vendor/sqlalchemy/orm/persistence.py", line 1697, in exec_
self._do_exec()
File "/home/seedit4me/calibre-web/vendor/sqlalchemy/orm/persistence.py", line 1930, in _do_exec
self._execute_stmt(delete_stmt)
File "/home/seedit4me/calibre-web/vendor/sqlalchemy/orm/persistence.py", line 1702, in _execute_stmt
self.result = self.query._execute_crud(stmt, self.mapper)
File "/home/seedit4me/calibre-web/vendor/sqlalchemy/orm/query.py", line 3568, in _execute_crud
return conn.execute(stmt, self._params)
File "/home/seedit4me/calibre-web/vendor/sqlalchemy/engine/base.py", line 1011, in execute
return meth(self, multiparams, params)
File "/home/seedit4me/calibre-web/vendor/sqlalchemy/sql/elements.py", line 298, in _execute_on_connection
return connection._execute_clauseelement(self, multiparams, params)
File "/home/seedit4me/calibre-web/vendor/sqlalchemy/engine/base.py", line 1130, in _execute_clauseelement
distilled_params,
File "/home/seedit4me/calibre-web/vendor/sqlalchemy/engine/base.py", line 1317, in _execute_context
e, statement, parameters, cursor, context
File "/home/seedit4me/calibre-web/vendor/sqlalchemy/engine/base.py", line 1511, in _handle_dbapi_exception
sqlalchemy_exception, with_traceback=exc_info[2], from_=e
File "/home/seedit4me/calibre-web/vendor/sqlalchemy/util/compat.py", line 182, in raise_
raise exception
File "/home/seedit4me/calibre-web/vendor/sqlalchemy/engine/base.py", line 1277, in _execute_context
cursor, statement, parameters, context
File "/home/seedit4me/calibre-web/vendor/sqlalchemy/engine/default.py", line 609, in do_execute
cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such module: fts5
[SQL: DELETE FROM books WHERE books.id = ?]
[parameters: (4,)]
(Background on this error at: http://sqlalche.me/e/13/e3q8)
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
File "/home/seedit4me/calibre-web/vendor/flask/app.py", line 2447, in wsgi_app
response = self.full_dispatch_request()
File "/home/seedit4me/calibre-web/vendor/flask/app.py", line 1952, in full_dispatch_request
rv = self.handle_user_exception(e)
File "/home/seedit4me/calibre-web/vendor/flask/app.py", line 1821, in handle_user_exception
reraise(exc_type, exc_value, tb)
File "/home/seedit4me/calibre-web/vendor/flask/_compat.py", line 39, in reraise
raise value
File "/home/seedit4me/calibre-web/vendor/flask/app.py", line 1950, in full_dispatch_request
rv = self.dispatch_request()
File "/home/seedit4me/calibre-web/vendor/flask/app.py", line 1936, in dispatch_request
return self.view_functions[rule.endpoint](**req.view_args)
File "/home/seedit4me/calibre-web/vendor/flask_login/utils.py", line 272, in decorated_view
return func(*args, **kwargs)
File "/home/seedit4me/calibre-web/cps/editbooks.py", line 212, in delete_book_ajax
return delete_book(book_id,book_format, False)
File "/home/seedit4me/calibre-web/cps/editbooks.py", line 287, in delete_book
log.debug_or_exception(e)
File "/home/seedit4me/calibre-web/cps/logger.py", line 50, in debug_or_exception
self.error(message, stacklevel=2, *args, **kwargs)
File "/usr/lib/python3.7/logging/__init__.py", line 1407, in error
self._log(ERROR, msg, args, **kwargs)
TypeError: _log() got an unexpected keyword argument 'stacklevel'

**Comment by OzzieIsaacs** [comment]
Looks like there is something wrong with the sqlalchemy installation:
`sqlite3.OperationalError: no such module: fts5`
(I have no idea why this happens). 
Could you please reinstall sqlalchemy? Which version of sqlalchemy are you using? (I hope not 1.4.x+ as it's not supported currently)
The 2nd error is indeed a problem of calibre-web. I'll fix this one, but I have another problem to fix first, so it takes some time...

**Comment by Oreolek** [comment]
Same error, different cause for me:

```
Traceback (most recent call last):
File "[calibre-web root]/vendor/sqlalchemy/engine/base.py", line 1277, in _execute_context
cursor, statement, parameters, context
File "[calibre-web root]/vendor/sqlalchemy/engine/default.py", line 609, in do_execute
cursor.execute(statement, parameters)
sqlite3.OperationalError: error in tokenizer constructor

The above exception was the direct cause of the following exception:
Traceback (most recent call last):
File "[calibre-web root]/cps/editbooks.py", line 281, in delete_book
calibre_db.session.query(db.Books).filter(db.Books.id == book_id).delete()
File "[calibre-web root]/vendor/sqlalchemy/orm/query.py", line 3926, in delete
delete_op.exec_()
File "[calibre-web root]/vendor/sqlalchemy/orm/persistence.py", line 1697, in exec_
self._do_exec()
File "[calibre-web root]/vendor/sqlalchemy/orm/persistence.py", line 1930, in _do_exec
self._execute_stmt(delete_stmt)
File "[calibre-web root]/vendor/sqlalchemy/orm/persistence.py", line 1702, in _execute_stmt
self.result = self.query._execute_crud(stmt, self.mapper)
File "[calibre-web root]/vendor/sqlalchemy/orm/query.py", line 3568, in _execute_crud
return conn.execute(stmt, self._params)
File "[calibre-web root]/vendor/sqlalchemy/engine/base.py", line 1011, in execute
return meth(self, multiparams, params)
File "[calibre-web root]/vendor/sqlalchemy/sql/elements.py", line 298, in _execute_on_connection
return connection._execute_clauseelement(self, multiparams, params)
File "[calibre-web root]/vendor/sqlalchemy/engine/base.py", line 1130, in _execute_clauseelement
distilled_params,
File "[calibre-web root]/vendor/sqlalchemy/engine/base.py", line 1317, in _execute_context
e, statement, parameters, cursor, context
File "[calibre-web root]/vendor/sqlalchemy/engine/base.py", line 1511, in _handle_dbapi_exception
sqlalchemy_exception, with_traceback=exc_info[2], from_=e
File "[calibre-web root]/vendor/sqlalchemy/util/compat.py", line 182, in raise_
raise exception
File "[calibre-web root]/vendor/sqlalchemy/engine/base.py", line 1277, in _execute_context
cursor, statement, parameters, context
File "[calibre-web root]/vendor/sqlalchemy/engine/default.py", line 609, in do_execute
cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) error in tokenizer constructor
[SQL: DELETE FROM books WHERE books.id = ?]
[parameters: (636,)]
(Background on this error at: http://sqlalche.me/e/13/e3q8)
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
File "[calibre-web root]/vendor/flask/app.py", line 2447, in wsgi_app
response = self.full_dispatch_request()
File "[calibre-web root]/vendor/flask/app.py", line 1952, in full_dispatch_request
rv = self.handle_user_exception(e)
File "[calibre-web root]/vendor/flask/app.py", line 1821, in handle_user_exception
reraise(exc_type, exc_value, tb)
File "[calibre-web root]/vendor/flask/_compat.py", line 39, in reraise
raise value
File "[calibre-web root]/vendor/flask/app.py", line 1950, in full_dispatch_request
rv = self.dispatch_request()
File "[calibre-web root]/vendor/flask/app.py", line 1936, in dispatch_request
return self.view_functions[rule.endpoint](**req.view_args)
File "[calibre-web root]/vendor/flask_login/utils.py", line 272, in decorated_view
return func(*args, **kwargs)
File "[calibre-web root]/cps/editbooks.py", line 212, in delete_book_ajax
return delete_book(book_id,book_format, False)
File "[calibre-web root]/cps/editbooks.py", line 287, in delete_book
log.debug_or_exception(e)
File "[calibre-web root]/cps/logger.py", line 50, in debug_or_exception
self.error(message, stacklevel=2, *args, **kwargs)
File "/usr/lib/python3.6/logging/__init__.py", line 1337, in error
self._log(ERROR, msg, args, **kwargs)
TypeError: _log() got an unexpected keyword argument 'stacklevel'
```
```

**Comment by OzzieIsaacs** [comment]
The error in the error is fixed, please update to the newest commit

**Comment by OzzieIsaacs** [comment]
@Oreolek: This is the solution for your problem: https://github.com/janeczku/calibre-web/issues/1639#issuecomment-733885878

**Comment by MarGer1962** [comment]
NAS-Synology DS716+ / 
DSM version: DSM 6.2.3-25426 Update 3
Installed linuxserver/calibre-web today using Docker

Made these directories on NAS:
Docker 
\Calibre \ books
>> in this directory i put the metadat.db  (Is this the correct place to put it there?)
>> in this directory i put the subdirectories\author \ ..epub
\Calibre \ config
(*) i attached log file which i found in \ config

At first i could start Calibre, the books where shown.
I changed admin and password > this was OK
I saw in the screen: "DB is not Writeable" (I did not understand that)
Then i went to my books

I could NOT change metadata
Then - when i wanted to select a book by searching on Author - i got this message

Question: what did i do wrong?
[calibre-web.log](https://github.com/janeczku/calibre-web/files/5845413/calibre-web.log)



*message
======================= begin of message
500 Internal Server Error: The server encountered an internal error and was unable to complete your request. Either the server is overloaded or there is an error in the application.
Traceback (most recent call last):
File "/app/calibre-web/cps/db.py", line 627, in fill_indexpage_with_archived_books
len(query.all()))
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/query.py", line 3373, in all
return list(self)
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/query.py", line 3535, in __iter__
return self._execute_and_instances(context)
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/query.py", line 3556, in _execute_and_instances
conn = self._get_bind_args(
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/query.py", line 3571, in _get_bind_args
return fn(
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/query.py", line 3550, in _connection_from_session
conn = self.session.connection(**kw)
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/session.py", line 1142, in connection
return self._connection_for_bind(
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/session.py", line 1150, in _connection_for_bind
return self.transaction._connection_for_bind(
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/session.py", line 409, in _connection_for_bind
self._assert_active()
File "/usr/local/lib/python3.8/dist-packages/sqlalchemy/orm/session.py", line 289, in _assert_active
raise sa_exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: This Session's transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely)
(sqlite3.OperationalError) attempt to write a readonly database
[SQL: UPDATE authors SET name=? WHERE authors.id = ?]
[parameters: (('Clancy, Tom', 13), ('Jager, Wiebe de', 14), ('Goyal, Kovid', 17), ('Ansdell, Gary', 18), ('Pavlicevic, Mercédès', 19), ('Schember, John', 20), ('Magee, Wendy', 21), ('Ramsey, David W.', 22) ... displaying 10 of 32 total bound parameter sets ... ('Burland, Karen', 45), ('Bache, Jane', 46))]
(Background on this error at: http://sqlalche.me/e/13/e3q8) (Background on this error at: http://sqlalche.me/e/13/7s2a)
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 2447, in wsgi_app
response = self.full_dispatch_request()
File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1952, in full_dispatch_request
rv = self.handle_user_exception(e)
File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1821, in handle_user_exception
reraise(exc_type, exc_value, tb)
File "/usr/local/lib/python3.8/dist-packages/flask/_compat.py", line 39, in reraise
raise value
File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1950, in full_dispatch_request
rv = self.dispatch_request()
File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1936, in dispatch_request
return self.view_functions[rule.endpoint](**req.view_args)
File "/app/calibre-web/cps/usermanagement.py", line 38, in decorated_view
return login_required(func)(*args, **kwargs)
File "/usr/local/lib/python3.8/dist-packages/flask_login/utils.py", line 272, in decorated_view
return func(*args, **kwargs)
File "/app/calibre-web/cps/web.py", line 718, in index
return render_books_list("newest", sort_param, 1, page)
File "/app/calibre-web/cps/web.py", line 421, in render_books_list
entries, random, pagination = calibre_db.fill_indexpage(page, 0, db.Books, True, order)
File "/app/calibre-web/cps/db.py", line 608, in fill_indexpage
return self.fill_indexpage_with_archived_books(page, pagesize, database, db_filter, order, False, *join)
File "/app/calibre-web/cps/db.py", line 630, in fill_indexpage_with_archived_books
logger.debug(e)
AttributeError: module 'cps.logger' has no attribute 'debug'
================== end of message

**Comment by OzzieIsaacs** [comment]
> I saw in the screen: "DB is not Writeable" (I did not understand that)
> I could NOT change metadata

The metadata.db file is not writeable (you can not do any changes to it). On editing metadata you put the database in the state where it wanted to change the data, but this could not be done. This causes all the later errors.

(And you found another line in the code where the debug logging isn't working)

**Comment by MarGer1962** [comment]
Hi Ozzie,
I am still working to fix the problem.
I think (but I am not an experienced user) that it has to do with some settings in NAS Synology (write/read/chage etc)
What I did for now was:

  *   I changed the right of metadata.db form readonly to write/readonly
I stopped calibre-web and started it again
That did not work

  *   Then: I changed the read/write rights for the directory calibre and also the subdirectories and all the files
I gave all the rights to the admin

I stopped calibre-web and started it again

Result: this seems to work, I can change metadata and the changed metadata of the book will be saved

  *   I have not figured out yet how to change the rights for other users in Calibre Web
  *   I succeeded in accessing Calibre-web from outside (I opened port 8083 in my router and this seems to work)



  *   Now I am trying to read an ebook in my browser without downloading



  *   Afterwards: testing for my mobile phone and my iPad


Thanks for your reply,
Question: do you want me to put my answers in GitHub? If so: how must I do that?
Greetings
Mark Gerrits
PS: Thanks a lot for your programm!

**Comment by OzzieIsaacs** [comment]
>  I have not figured out yet how to change the rights for other users in Calibre Web

Just go to the admin section, click on the users name and you will see a page similar to the user page you see be clicking on your user name, there you can change all setting you'd like.
![grafik](https://user-images.githubusercontent.com/6438895/105404183-f80dca80-5c29-11eb-8c02-8322cabf9765.png)
You can edit the default settings for new users in admin section -> Edit UI configuration -> 
![grafik](https://user-images.githubusercontent.com/6438895/105404303-196eb680-5c2a-11eb-8993-16b78a866b74.png)

> Question: do you want me to put my answers in GitHub? If so: how must I do that?

You already did by answering the email, so be careful what you write, the whole world is listening...

**Comment by OzzieIsaacs** [comment]
No response for longer time, so I guess the problem is solved

**Comment by MarGer1962** [comment]
Hi Ozzie,
I'm sorry that i did not got back to you earlier (had some private problems to deal with). I followed your advice how to adjust an extra user. This works.
I also can now read books from phone or ipad if i am on my way (that is: not connected to my home network.)

The problem i faced (no synchronisation when i made changes in books or metadata) is due to the way Calibre and Calibre-Web use the metadata.db file.
There is already a topic written about that (?), in which you direct the users of your program to the Calibre website for a solution.
I solved that problem for myself, and it's working fine now. 
Is there - by your knowledge - anyone on your platform interested? 
If so: let me know.
Greetz, Mark
